### PR TITLE
WIP: Parse MDX

### PR DESCRIPTION
**Pull request by valentinnodan**
Only able to parse one line export-statements

### CODE REVIEW

File: src/main/java/org/intellij/plugin/mdx/lang/MdxFileType.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import com.intellij.openapi.fileTypes.LanguageFileType;
+import org.intellij.plugin.mdx.ide.MDXIcons;
+import org.jetbrains.annotations.Nls;
+import org.jetbrains.annotations.NonNls;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
+
+public class MdxFileType extends LanguageFileType {
```
Comment (anstarovoyt) [inline src/main/java/org/intellij/plugin/mdx/lang/MdxFileType.java:12]: why java? :)

File: src/main/java/org/intellij/plugin/mdx/lang/MdxLanguage.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package org.intellij.plugin.mdx.lang;
+
+import com.intellij.lang.Language;
+import com.intellij.psi.templateLanguages.TemplateLanguage;
+import org.intellij.plugins.markdown.lang.MarkdownLanguage;
+
+public class MdxLanguage extends Language implements TemplateLanguage {
```
Comment (anstarovoyt) [inline src/main/java/org/intellij/plugin/mdx/lang/MdxLanguage.java:7]: why java?

File: src/main/java/org/intellij/plugin/mdx/lang/psi/MdxFileViewProvider.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            }
+        }
+
+        val parserDefinition = LanguageParserDefinitions.INSTANCE.forLanguage(lang) ?: return null
+
+        val psiFile = parserDefinition.createFile(this)
+        if (lang === templateDataLanguage && psiFile is PsiFileImpl) {
+            val debugName = lang.displayName.toUpperCase().replace(' ', '_')
+            val mdxTemplate =
+                    templateDataElementType(debugName)
```
Comment (anstarovoyt) [inline src/main/java/org/intellij/plugin/mdx/lang/psi/MdxFileViewProvider.kt:39]: it is always the same :)

File: src/main/java/org/intellij/plugin/mdx/lang/psi/MdxFile.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package org.intellij.plugin.mdx.lang.psi;
+
+import com.intellij.extapi.psi.PsiFileBase;
+import com.intellij.openapi.fileTypes.FileType;
+import com.intellij.psi.FileViewProvider;
+import org.intellij.plugin.mdx.lang.MdxFileType;
+import org.intellij.plugin.mdx.lang.MdxLanguage;
+import org.jetbrains.annotations.NotNull;
+
+public class MdxFile extends PsiFileBase {
```
Comment (anstarovoyt) [inline src/main/java/org/intellij/plugin/mdx/lang/psi/MdxFile.java:10]: why java

File: src/main/java/org/intellij/plugin/mdx/lang/parse/MdxPsiBuilderFillingVisitor.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    private fun ensureBuilderInPosition(position: Int) {
+        //Advance to position and maybe skip whitespaces
+        while (builder.currentOffset < position) {
+            builder.advanceLexer()
+        }
+    }
+
+    fun leadingWhitespaces(): WhitespacesAndCommentsBinder? {
```
Comment (anstarovoyt) [inline src/main/java/org/intellij/plugin/mdx/lang/parse/MdxPsiBuilderFillingVisitor.kt:58]: hmmm

File: src/main/java/org/intellij/plugin/mdx/lang/parse/MdxLexer.kt [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+ *
+ * @see org.intellij.plugins.markdown.lang.lexer.MarkdownToplevelLexer
+ */
+
+class MdxLexer : MarkdownToplevelLexer(MdxFlavourDescriptor) {
+    private var myBuffer: CharSequence? = null
+    private var myBufferStart = 0
+    private var myBufferEnd = 0
+
+    private var myLexemes: ArrayList<IElementType>? = null
```
Comment (anstarovoyt) [inline src/main/java/org/intellij/plugin/mdx/lang/parse/MdxLexer.kt:21]: MutableList

File: src/main/java/org/intellij/plugin/mdx/lang/psi/MdxFileViewProvider.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import gnu.trove.THashSet
+import org.intellij.plugin.mdx.lang.MdxLanguage
+import org.intellij.plugin.mdx.lang.parse.MdxFlavourDescriptor
+import org.intellij.plugins.markdown.lang.parser.MarkdownParserManager
+
+
+class MdxFileViewProvider(manager: PsiManager, virtualFile: VirtualFile, eventSystemEnabled: Boolean)
+    : MultiplePsiFilesPerDocumentFileViewProvider(manager, virtualFile, eventSystemEnabled), TemplateLanguageFileViewProvider {
+
+    private val myRelevantLanguages = THashSet<Language>()
```
Comment (anstarovoyt) [inline src/main/java/org/intellij/plugin/mdx/lang/psi/MdxFileViewProvider.kt:18]: can be mutableSetOf(baseLanguage, templateDataLanguage)

File: src/main/java/org/intellij/plugin/mdx/highlighting/MdxTemplateHighlighter.kt [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import com.intellij.openapi.editor.ex.util.LayeredLexerEditorHighlighter
+import com.intellij.openapi.fileTypes.FileType
+import com.intellij.openapi.fileTypes.SyntaxHighlighterFactory
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import org.intellij.plugin.mdx.lang.parse.MdxTokenTypes
+import org.intellij.plugins.markdown.lang.MarkdownElementType
+import org.intellij.plugins.markdown.lang.MarkdownTokenTypes
+
+class MdxTemplateHighlighter(project: Project?, virtualFile: VirtualFile?, colors: EditorColorsScheme)
```
Comment (anstarovoyt) [inline src/main/java/org/intellij/plugin/mdx/highlighting/MdxTemplateHighlighter.kt:17]: why Template? I believe it was incorrect copy-paste from my side. Let's choose a better name

File: src/main/java/org/intellij/plugin/mdx/lang/MdxFileType.kt [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+import com.intellij.openapi.fileTypes.LanguageFileType
+import org.intellij.plugin.mdx.ide.MDXIcons
+import org.jetbrains.annotations.Nls
+import org.jetbrains.annotations.NonNls
+import javax.swing.Icon
+
+class MdxFileType private constructor() : LanguageFileType(MdxLanguage) {
+    override fun getName(): String {
+        return "MDX File"
```
Comment (anstarovoyt) [inline src/main/java/org/intellij/plugin/mdx/lang/MdxFileType.kt:11]: no "File" just "MDX" and the same thing for description

File: src/main/java/org/intellij/plugin/mdx/highlighting/MdxTemplateHighlighter.kt (hunk 8 above)
Comment (valentinnodan) [reply src/main/java/org/intellij/plugin/mdx/highlighting/MdxTemplateHighlighter.kt:17]: Maybe just MdxHighlighter better?

Comment (anstarovoyt) [reply src/main/java/org/intellij/plugin/mdx/highlighting/MdxTemplateHighlighter.kt:17]: MdxEditorHighlighter

**Pull request merged**